src/games/seabattle/ai.ts

```diff
@@ -0,0 +1,189 @@
+  }
+
+  makeSalvoMove(cell: ICell, playerID: string) {
+    return { action: { type: 'MAKE_MOVE', payload: { type: 'salvo', args: [cell.x, cell.y], playerID } } };
+  }
+
+  generateMove(playerID: string, state: IPlayState) {
+    const salvos = state.G.salvos.filter((salvo: ISalvo) => (
+      salvo.player === Number(playerID) &&
+      salvo.hit === true &&
+      typeof salvo.hitShip !== 'undefined'));
+    for (const salvo of salvos) {
+      // if we have already fired salvos, see if we've hit any ships that remain unsunk
+      if (!this.isShipSunk(state, salvo.hitShip)) {
+        const otherHitSalvos = this.getOtherSalvosHitShip(state, salvo.hitShip);
+        if (otherHitSalvos.length >= 2) {
+          // we have hit this ship at least twice, so we can make an intelligent move
+          return this.getNextShipFoundMove(state, otherHitSalvos);
+        }
+        // no other salvos for the same ship were hit, so hit a random neighbor
+        return this.getRandomNeighbor(state, salvo);
+      }
+    }
+    // generate a random move
+    return this.generateRandomMove(state);
+  }
+
+  getNextShipFoundMove(state: IPlayState, hitSalvos: ISalvo[]): ICell {
+    const xMap = hitSalvos.map((salvo) => salvo.cell.x);
+    const yMap = hitSalvos.map((salvo) => salvo.cell.y);
+
+    const minSalvoPos = {
+      x: Math.min(...xMap),
+      y: Math.min(...yMap),
+    };
+    const maxSalvoPos = {
+      x: Math.max(...xMap),
+      y: Math.max(...yMap),
+    };
+    // either x or y must be 0, given that the ship must be horizontal or vertical
+    const diffMinMaxSalvoPos = {
+      x: minSalvoPos.x - maxSalvoPos.x,
+      y: minSalvoPos.y - maxSalvoPos.y,
+    };
+    const direction = diffMinMaxSalvoPos.x === 0 ? { x: 0, y: 1 } : { x: 1, y: 0 };
+    const diffLength = diffMinMaxSalvoPos.x + diffMinMaxSalvoPos.y;
+    if (diffLength === hitSalvos.length) {
+      // This means that there is no "hole" in the salvos, therefore we must try the edges
+      const possibleMoves: ICell[] = [
+        { x: minSalvoPos.x - direction.x, y: minSalvoPos.y - direction.y },
+        { x: maxSalvoPos.x + direction.x, y: maxSalvoPos.y + direction.y },
+      ];
+      return this.anyValidMove(state, possibleMoves);
+    } else {
+      return this.anyValidMove(state, this.allCellsBetween(minSalvoPos, maxSalvoPos, direction, diffLength));
+      // return anyValidMove([ all cells between minSalvoPos and maxSalvoPos ]);
+    }
+  }
+
+  allCellsBetween(min: ICell, max: ICell, direction: ICell, diffLength: number) {
+    const cells: ICell[] = [];
+    const absDiffLength = Math.abs(diffLength);
+    if (direction.x) {
+      for (let x = min.x - absDiffLength; x <= max.x + absDiffLength; x++) {
+        cells.push({ x: x, y: max.y });
+      }
+    } else {
+      for (let y = min.y - absDiffLength; y <= max.y + absDiffLength; y++) {
+        cells.push({ x: max.x, y: y });
+      }
+    }
```

Comment:
You definitely dont need two for loops. You can start with the min and add both x and y of direction to the min in each iteration, so you only have one for loop.